master/buildbot/data/buildrequests.py

```diff
@@ -62,6 +64,31 @@ def startConsuming(self, callback, options, kwargs):
         return self.master.mq.startConsuming(callback,
                                              ('buildrequests', buildrequestid, None))
 
+    @defer.inlineCallbacks
+    def control(self, action, args, kwargs):
+        if action != "cancel":
+            raise ValueError("action: {} is not supported".format(action))
+        brid = kwargs['buildrequestid']
+        # first, try to claim the request; if this fails, then it's too late to
+        # cancel the build anyway
+        try:
+            b = yield self.master.db.buildrequests.claimBuildRequests(brids=[brid])
+        except AlreadyClaimedError:
+            builds = yield self.master.data.get(("buildrequests", brid, "builds"))
```

Comment:
There is a race condition here. The buildrequest might be claimed, but the build are not yet created. The claiming master has to find a slave, substanciate it, etc before creating the build.
As ee don't have a proper solution for this race condition, I would at least create a comment here stating this, and create a trac to follow up on this problem